dropshot/examples/multiple-servers.rs

```diff
@@ -0,0 +1,377 @@
+     *
+     * If we receive a new server on `running_servers_rx`, we added it to
+     * `running_servers`. If `running_servers_rx` indicates the channel is
+     * closed, we know all server contexts have been dropped and no servers
+     * remain.
+     */
+    loop {
+        tokio::select! {
+            maybe_new_server = running_servers_rx.recv() => {
+                match maybe_new_server {
+                    Some(server) => running_servers.push(server),
+                    None => return Ok(()),
+                }
+            }
+
+            maybe_result = running_servers.next() => {
+                if let Some(result) = maybe_result {
+                    // Fail if any server failed to shut down.
+                    result?;
+                }
+            }
+        }
+    }
+}
+
+type ServerShutdownFuture = BoxFuture<'static, Result<(), String>>;
+
+/**
+ * Application-specific server context (state shared by handler functions)
+ */
+struct MultiServerContext {
+    // All running servers have the same underlying `shared` context.
+    shared: Arc<SharedMultiServerContext>,
+
+    // `name` and `log` are unique to each running server, allowing them to log
+    // their own name when they shut down.
+    name: String,
+    log: Logger,
+}
+
+impl Drop for MultiServerContext {
+    fn drop(&mut self) {
+        info!(self.log, "shut down server {:?}", self.name);
+    }
+}
+
+/**
+ * Context shared by all running servers.
+ */
+struct SharedMultiServerContext {
+    servers: Mutex<HashMap<String, HttpServer<MultiServerContext>>>,
+    started_server_shutdown_handles: mpsc::Sender<ServerShutdownFuture>,
+}
+
+impl SharedMultiServerContext {
+    fn new(
+        started_server_shutdown_handles: mpsc::Sender<ServerShutdownFuture>,
+    ) -> Self {
+        Self { servers: Mutex::default(), started_server_shutdown_handles }
+    }
+
+    async fn start_server(
+        self: &Arc<Self>,
+        name: &str,
+        bind_address: SocketAddr,
+    ) -> Result<(), String> {
+        let mut servers = self.servers.lock().await;
+        let slot = match servers.entry(name.to_string()) {
+            Entry::Occupied(_) => {
+                return Err(format!("already running a server named {name:?}",))
+            }
+            Entry::Vacant(slot) => slot,
+        };
+
+        /*
+         * For simplicity, we'll configure an "info"-level logger that writes to
+         * stderr assuming that it's a terminal.
+         */
+        let config_logging =
+            ConfigLogging::StderrTerminal { level: ConfigLoggingLevel::Info };
+        let log = config_logging
+            .to_logger(format!("example-multiserver-{name}"))
+            .map_err(|error| format!("failed to create logger: {}", error))?;
+
+        /*
+         * Build a description of the API.
+         *
+         * TODO: Could `ApiDescription` implement `Clone`, or could we pass an
+         * `Arc<ApiDescription>` instead?
```

Comment:
This seems like a good idea.

Reply:
Implementing `Clone` for `ApiDescription` is doable but kinda gnarly; it requires a bunch of manual `impl Clone for ...` to avoid requiring the server contexts to be cloneable (which is not necessary). OTOH, `Arc<ApiDescription>` seems problematic because of [ApiDescription::into_router()](https://github.com/oxidecomputer/dropshot/blob/01263bc1e7074bae8f6eee063d9a34981949b369/dropshot/src/api_description.rs#L897-L899). Maybe `ApiDescription` could hold two `Arc<_>`s as its fields (making it easily cloneable), and `into_router()` could return an `Arc<HttpRouter<Context>>`? I haven't looked at this very much - do you want me to poke at it more before merging this?

Reply:
Definitely not a blocker on this!